tower-http/src/catch_panic.rs

```diff
@@ -0,0 +1,202 @@
+    pub struct ResponseFuture<F> {
+        #[pin]
+        kind: Kind<F>,
+    }
+}
+
+pin_project! {
+    #[project = KindProj]
+    enum Kind<F> {
+        Panicked {
+            panic_err: Option<Box<dyn Any + Send + 'static>>,
+        },
+        Future {
+            #[pin]
+            future: CatchUnwind<AssertUnwindSafe<F>>,
+        }
+    }
+}
+
+impl<F, ResBody, E> Future for ResponseFuture<F>
+where
+    F: Future<Output = Result<Response<ResBody>, E>>,
+    ResBody: Body<Data = Bytes> + Send + 'static,
+{
+    type Output = Result<Response<UnsyncBoxBody<Bytes, ResBody::Error>>, E>;
+
+    #[allow(warnings)]
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.project().kind.project() {
+            KindProj::Panicked { panic_err } => Poll::Ready(Ok(panic_err_to_response(
+                panic_err.take().expect("future polled after completion"),
+            ))),
+            KindProj::Future { future } => match ready!(future.poll(cx)) {
+                Ok(Ok(res)) => Poll::Ready(Ok(res.map(|body| body.boxed_unsync()))),
+                Ok(Err(svc_err)) => Poll::Ready(Err(svc_err)),
+                Err(panic_err) => Poll::Ready(Ok(panic_err_to_response(panic_err))),
+            },
+        }
+    }
+}
+
+fn panic_err_to_response<E>(
+    err: Box<dyn Any + Send + 'static>,
+) -> Response<UnsyncBoxBody<Bytes, E>> {
+    let message: Cow<_> = if let Some(s) = err.downcast_ref::<String>() {
+        s.to_owned().into()
+    } else if let Some(s) = err.downcast_ref::<&str>() {
+        (*s).into()
+    } else {
+        "`CatchPanic` middleware was unable to obtain panic info".into()
+    };
+    let body = format!("Service panicked: {}", message);
+
+    let mut res = Response::new(Full::from(body).map_err(|err| match err {}).boxed_unsync());
+    *res.status_mut() = StatusCode::INTERNAL_SERVER_ERROR;
+
+    res
+}
+
+#[cfg(test)]
+mod tests {
+    #![allow(unreachable_code)]
+
+    use super::*;
+    use hyper::{Body, Response};
+    use std::convert::Infallible;
+    use tower::{ServiceBuilder, ServiceExt};
+
+    #[tokio::test]
+    async fn panic_before_returning_future() {
+        let svc = ServiceBuilder::new()
+            .layer(CatchPanicLayer::new())
+            .service_fn(|_: Request<Body>| {
+                panic!("service panic");
+                async { Ok::<_, Infallible>(Response::new(Body::empty())) }
+            });
+
+        let req = Request::new(Body::empty());
+
+        let res = svc.oneshot(req).await.unwrap();
+
+        assert_eq!(res.status(), StatusCode::INTERNAL_SERVER_ERROR);
+        let body = hyper::body::to_bytes(res).await.unwrap();
+        assert_eq!(&body[..], b"Service panicked: service panic");
+    }
+
+    #[tokio::test]
+    async fn panic_in_future() {
+        let svc = ServiceBuilder::new()
+            .layer(CatchPanicLayer::new())
+            .service_fn(|_: Request<Body>| async {
+                panic!("future panic");
```

Comment:
One of the tests should panic with a format string, so that both panic payload cases (`&str` and `String`) are tested.